### ISSUE TITLE
There is a noticeable delay until the Snooze Tabs dropdown appears

### ISSUE THREAD

**Issue by SoftVision-CiprianMuresan** [bwinton/SnoozeTabs#79]
[Affected versions]:
- Nightly 53.0a1
- Snooze Tabs 1.0 (built on 1/9/2016)

[Affected Platforms]:
- All Windows
- All Mac
- All Linux

[Prerequisites]:
- Have a Firefox profile with the latest Snooze Tabs add-on version (1.0-dev, built on 1/9/2016) loaded as a temporary add-on.

[Steps to reproduce]:
1. Focus the window with the profile from prerequisites.
2. Click the Snooze Tabs button from the browser toolbar.
3. Observe the time it takes for the dropdown to appear.

[Expected results]:
- The dropdown appears instantly.

[Actual results]:
- There is a noticeable delay until the dropdown appears.

[Notes]:
- Here is a screen recording with the issue:
![there is a noticeable delay until the snooze tabs dropdown appears](https://cloud.githubusercontent.com/assets/18304511/21770404/c6d07d3c-d68b-11e6-8b3a-5c2365522656.gif)

**Comment by bwinton** [comment]
So, I dug into this a little, and it looks like it takes between 600-900ms to go through the `snoozetabs-content.js`'s `init()` method, which seems kinda long.  There's more delay before the `init` even gets called, but I don't know how to measure that, much less fix it…

@lmorchard: What do you think about rendering the html in the `background.js`, and sending it to the panel on open?  It doesn't seem to be particularly dependent on which url the tab is showing (aside from some special urls that only get the manage pane)…  Actually, what about pre-rendering both states, and sending the appropriate one when the panel opens?

**Comment by lmorchard** [comment]
Could even just manually code the initial main panel HTML in `src/popup/snooze.html` - React will swoop in after and apply whatever changes to the DOM it needs. Probably easier than trying to get things to render in the background script. Not sure if that will speed up time to when it's actually ready for interaction, though, since it's only after the first render that click handlers are active. But, it might at least *feel* more ready

**Comment by bwinton** [comment]
I think having it appear sooner would be a plus, even if it wasn't live.

**Comment by lmorchard** [comment]
FWIW, [experimenting in a branch](https://github.com/lmorchard/SnoozeTabs/commit/cf43d048890e5112498046bc0517ec46d6ed4157), but not really seeing much in the way of speed-up. I grabbed HTML source from the Browser Toolbox, post-render, and pasted it into `src/popup/snooze.html`. Also minified the JS and tossed a `defer` on the script tag. Tossed in some timers for the heck of it, too.
```
init: timer started
render: timer started
render: 184.22ms
init: 191.27ms
```
I'm thinking the delay isn't in the code, but comes in the setup & load of the panel itself. Trying to see if there's anything to wrangle there.

**Comment by lmorchard** [comment]
Oh, also tried commenting out all the setup code in `init()` and just went with the copypasta'd pre-rendered HTML. Still seems to be delayed.

**Comment by bwinton** [comment]
Yeah, there's some delay there I think @kmaglione might have some insights on what's happening, and how we could fix it...

**Comment by lmorchard** [comment]
I *did* just manage to get a speed up by commenting out the `<script>` tag altogether. Maybe we need a smaller stub script in there that sets everything up? Ugh

**Comment by lmorchard** [comment]
Also trying to see if/how to reduce the JS size itself in case parsing & etc is the trouble here - not entirely sure why it's so large.

**Comment by bwinton** [comment]
I mean, it _is_ loading in a 1.2M `snooze-content.js` file every time it opens, right?  That's gotta take some time.

**Comment by lmorchard** [comment]
Well, it's 486k minified. But, still larger than I'd like.

**Comment by bwinton** [comment]
Which brings me back to the "put the react in background.js, which gets loaded once, and push the html to the content script when we need to...  :wink:

**Comment by lmorchard** [comment]
Not really sure how to do that, since React works with the DOM content & events.

**Comment by bwinton** [comment]
Hmm.  Yeah.  Perhaps we could pull some of the larger stuff out (like the i18n and moment locales)?

**Comment by lmorchard** [comment]
That stuff isn't super big. I think most of the bulk is from the Babel ES6 polyfills and the calendar / time components.

Also tried tossing in a little bootstrap script that would load the main script after the panel loaded. But, ugh, ran into a CSP error `Content Security Policy: The page’s settings blocked the loading of a resource at self (“script-src moz-extension://9fbdc1af-6499-594a-9981-68cb00ffa61e”). Source: 
    var script = document.createElement....`

**Comment by lmorchard** [comment]
I did `more dist/popup/snooze-popup.js` and leaned on the space bar. Looked like this:
```
➜  SnoozeTabs git:(79-dropdown-delay) ✗ du -hs node_modules/babel-polyfill/dist/polyfill.min.js node_modules/rc-*/lib
 96K	node_modules/babel-polyfill/dist/polyfill.min.js
 16K	node_modules/rc-align/lib
 28K	node_modules/rc-animate/lib
208K	node_modules/rc-calendar/lib
 44K	node_modules/rc-time-picker/lib
 40K	node_modules/rc-trigger/lib
 60K	node_modules/rc-util/lib
```

**Comment by lmorchard** [comment]
Trying [this ugly bootstrap hack to defer loading of `snooze-content.js`](https://github.com/lmorchard/SnoozeTabs/commit/651b2980ca20fdbc404fba0d6fd9295b3a33953e), but it doesn't seem to help much to speed up the initial panel appearance.

**Comment by lmorchard** [comment]
Still poking around at ways to get a faster startup. For bundle size, I found [`rollup-analyze`](https://www.npmjs.com/package/rollup-analyzer) and did [a breakdown on the popup bundle before minification](https://gist.github.com/lmorchard/009d2829d0eef0c84bd5c6b86c72ca08). 

TL;DR: moment.js takes up around 300kb, react takes up a chunk, calendar component another chunk.

**Comment by bwinton** [comment]
But I thought I heard that the size of the script didn't matter that much…  Like, _any_ script would cause a pretty big delay.

**Comment by lmorchard** [comment]
Yeah, I'm noticing a decent delay with any script, but also wondering if the size makes that delay even longer

**Comment by lmorchard** [comment]
Okay, now I'm thoroughly confused on this issue. I've got [this branch](https://github.com/lmorchard/SnoozeTabs/tree/79-dropdown-delay) I've been working through. Can anyone else see what it does on another machine? (e.g. @bwinton / @sevaan) My CircleCI run has builds as of a few minutes ago: 

https://115-74497025-gh.circle-artifacts.com/0/tmp/circle-artifacts.piOmiak/snoozetabs-1.0.15.zip 
https://115-74497025-gh.circle-artifacts.com/0/tmp/circle-artifacts.piOmiak/snoozetabs-dev.zip

In this build, I'm trying a small bootstrap script that slightly delays loading the main React UI in the hopes that that will yield to allow the panel to have an initial render. Kind of trying voodoo at this point.

I don't usually see a delay with a clean profile on release Firefox 51.0.1 for OS X and this build loaded as a temporary add-on via about:debugging. The panel immediately sprouts from the button, sometimes stutters a little mid-way as things render. Occasionally, it will pause and I'll see the initial static HTML with '...' placeholders I've stuck in for dates. 

I've mostly been trying this stuff on Nightly - but Nightly's lately been stalling & beach balling all over the place for me. That's why I didn't necessarily think I could reproduce this issue at first - because *everything* has significant delays and it didn't seem unusual.

So, I'm not sure what I'm seeing at this point

**Comment by lmorchard** [comment]
FWIW, it could also be that this bootstrap script has fixed this issue, but I need a sanity check because I think things are going weird on my machine.

**Comment by chuckharmston** [comment]
Are there some perceived performance tricks you can do here? For example, giving the body a minimum height should at least give the popup some body before React runs through everything.

**Comment by lmorchard** [comment]
> Are there some perceived performance tricks you can do here?

Yup, already tried some things like that. Currently, on this branch, I have a bit of HTML pre-rendered statically into the initial `snooze.html` along with the CSS and a bootstrap script that waits 10ms before injecting a script tag with the rest of the JS bundle.

**Comment by ckprice** [comment]
> FWIW, it could also be that this bootstrap script has fixed this issue, but I need a sanity check because I think things are going weird on my machine.

Gave https://github.com/lmorchard/SnoozeTabs/tree/79-dropdown-delay a try side by side with the current production version.

Video: https://drive.google.com/file/d/0B5T1SCKaYZL8SXFySW5PR1h2NzQ/view?usp=sharing

Seems like there's an incremental improvement over the production version. This new version (bottom) seems to put something on the screen faster, but takes a bit before it loads the entire panel.

**Comment by ckprice** [comment]
Made some incremental changes, but to truly fix this would require a re-write. Moving out.

**Comment by lmorchard** [comment]
Just as an update on further tinkering I've been doing: Assuming JS bundle size is a big contributor, I started looking into ways to reduce it. 

* Added a bootstrap script to the page that defers loading the big JS blob until after the HTML has had an initial render - this at least lets the popup start to expand. That means the delay now appears in "..." placeholders for dates, an initial flash of English strings in non-en-US locales, and "Real Soon Now" popping in after load on dev builds.

* Tried Webpack, but it produced roughly the same sized JS bundle as Rollup. 

* Used [`rollup-analyzer`](https://github.com/doesdev/rollup-analyzer/) and [`webpack-bundle-analyzer`](https://github.com/th0r/webpack-bundle-analyzer). Both of these told me that React, Moment, and the rc-calendar widget are just big hunks of code. Unfortunately, those things are the core of the product right now. Excluding locale modules that Moment ships can shave 100k or so off the size, but `react-dom` is the real big slab.

    <img width="1746" alt="screenshot 2017-03-14 14 14 56" src="https://cloud.githubusercontent.com/assets/21687/23915883/c3c3d7f8-08c0-11e7-8043-9f3b1c027b2c.png">

The "rewrite" part is the next best terrible option I can come up with: Maybe we can defer loading React until a calendar view is summoned, and use vanilla JS for the "main" panel with snooze time options. Simple lists of clickable options are relatively easy. Calendar widgets are hard.

**Comment by lmorchard** [comment]
Just because I can't seem to stop beating up on this zombie horse: I found a lighterweight React alternative called Preact [with pretty simple instructions for switching over](https://preactjs.com/guide/switching-to-preact#easy-preact-compat-alias). So, I tried that.

Good news: It reduced the bundle by ~30% and the UI still works! 😄 

Bad news: It didn't seem to have much influence on init speed for the popup UI. 😞 

I stuck a `console.time()` in `snooze-bootstrap.js` and a `console.timeEnd()` in `snooze-popup.js`. The results end up all over the place, but about in the same range with React or Preact. 

For example, opening & closing the popup a few times, here are some load times with a 383kb bundle using Preact:
```
fe_load: 434.29ms
fe_load: 367.52ms
fe_load: 752.52ms
fe_load: 398.98ms
fe_load: 5152.78ms
fe_load: 435.6ms
```
And with a 508kb bundle using React:
```
fe_load: 474.45ms
fe_load: 679.03ms
fe_load: 343.98ms
fe_load: 3957.38ms
fe_load: 430.01ms
fe_load: 406.77ms
```
So, roughly around 400-700ms with occasional spikes up to several whole seconds. Haven't yet worked out what's causing those spikes.

And just for good measure, here's what the treemap looks like with Preact:
<img width="1746" alt="screenshot 2017-03-17 15 12 26" src="https://cloud.githubusercontent.com/assets/21687/24059140/41343e2a-0b24-11e7-8df8-a1f987165e31.png">

**Comment by lmorchard** [comment]
FWIW, I'm just going to kick this out of the milestone: I don't have a good fix in mind, but we *might* have some general WebExtension perf improvements coming up

**Comment by lmorchard** [comment]
Going to assert I've banged my head on this long enough to exhaust most options short of a) drastically reworking things with regard to React & bundling or b) addressing issues upstream in WebExtensions. 

I don't think these activities entirely serve the purposes of SnoozeTabs as an experiment in Test Pilot. If we graduate to life as an independent add-on, this could all be worth revisiting.

**Comment by mfn** [comment]
I don't understand why this is closed when it's clearly not fixed?

@lmorchard no one expects you to fix it, but IMHO it should be left open so everyone is easily aware of this defect.

**Comment by bwinton** [comment]
@mfn: It's because it's unfixable by this add-on, and the issue list isn't the place to list all the random things people don't like about the current version of the add-on.  Maybe it'll get fixed by a future version of the WebExtension API, or maybe not, but this isn't the place to track that.
